cognite/neat/rules/transformers/_base.py

```diff
@@ -1,24 +1,83 @@
 from abc import ABC, abstractmethod
 from collections.abc import MutableSequence
-from typing import Generic, TypeVar
+from dataclasses import dataclass
+from typing import Any, Generic, TypeVar
 
-from cognite.neat.rules._shared import Rules
+from cognite.neat.issues import IssueList
+from cognite.neat.issues.errors import NeatTypeError, NeatValueError
+from cognite.neat.rules._shared import InputRules, Rules, T_Rules, VerifiedRules
 
 T_RulesIn = TypeVar("T_RulesIn", bound=Rules)
 T_RulesOut = TypeVar("T_RulesOut", bound=Rules)
 
 
+@dataclass
+class OutRules(Generic[T_Rules], ABC):
+    """This is a base class for all rule states."""
+
+
+@dataclass
+class JustRules(OutRules[T_Rules]):
+    """This represents a rule that exists"""
+
+    rules: T_Rules
+
+
+@dataclass
+class MaybeRules(OutRules[T_Rules]):
+    """This represents a rule that may or may not exist"""
+
+    rules: T_Rules | None
+    issues: IssueList
+
+
+@dataclass
+class ReadRules(OutRules[T_Rules]):
+    """This represents a rule that does not exist"""
+
+    rules: T_Rules
+    read_context: dict[str, Any]
+
+
 class RulesTransformer(ABC, Generic[T_RulesIn, T_RulesOut]):
-    """This is the base class for all rule transformers."""
+    """This is the base class for all rule transformers.
+
+    Note transformers follow the functional pattern Monad
+    https://en.wikipedia.org/wiki/Monad_(functional_programming)
+    """
 
     @abstractmethod
-    def transform(self, rules: T_RulesIn) -> T_RulesOut:
+    def transform(self, rules: T_RulesIn | OutRules[T_RulesIn]) -> OutRules[T_RulesOut]:
         """Transform the input rules into the output rules."""
         raise NotImplementedError()
 
+    @classmethod
+    def _to_rules(cls, rules: T_RulesIn | OutRules[T_RulesIn]) -> T_RulesIn:
+        if isinstance(rules, JustRules):
+            return rules.rules
+        elif isinstance(rules, MaybeRules):
```

Comment:
Would we have extra information why mayberules are None?
Something to discuss on Wed.